### PR TITLE
[Bug fix] expm1_bw: take exp(x) directly instead of expm1(x) + 1

### LINKED ISSUE #55723: [Bug] ttnn.expm1_bw returns a zero gradient for x <= -6.25 (bf16) / -17 (fp32): the derivative is built as expm1(x) + 1
### Describe the bug

`ttnn.expm1_bw` returns a gradient of exactly `0` for every input below `-6.25` in bfloat16 and `-17` in float32, although `d/dx expm1(x) = exp(x)` is an ordinary normal float all the way down to `x = -87.34`, where it is still `1.2e-37`.

The composite builds the derivative out of the forward result instead of calling `exp`
([`unary_backward.cpp#L1334-L1343`](https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1334-L1343)):

```cpp
// bw(expm1) = grad * expm1(input) + 1
std::vector<Tensor> expm1_bw(
    const Tensor& grad, const Tensor& input, const std::optional<MemoryConfig>& output_mem_config) {
    std::vector<Tensor> grad_tensor;
    Tensor eresult = ttnn::expm1(input, output_mem_config);
    Tensor rp1 = ttnn::add(eresult, 1.0f, std::nullopt, output_mem_config);
    Tensor result = ttnn::multiply(grad, rp1, std::nullopt, output_mem_config);
    grad_tensor.emplace_back(result);
    return grad_tensor;
}
```

`expm1(x) + 1` equals `exp(x)` in exact arithmetic only. `expm1(x)` tends to `-1` from above and rounds to **exactly** `-1` as soon as `exp(x)` falls below half an ULP of `1` — below `2^-9` for bfloat16 and `2^-25` for float32. The `+ 1` then cancels completely and the entire gradient is lost. This is the cancellation `expm1` exists to avoid, reintroduced in the one place where plain `exp` was what was wanted.

### Repro

```python
import torch, ttnn

device = ttnn.open_device(device_id=0)
x = torch.tensor([[-5.0, -10.0, -17.0, -20.0, -30.0, -60.0, -85.0]], dtype=torch.float32)
t = ttnn.from_torch(x, dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)
g = ttnn.from_torch(torch.ones_like(x), dtype=ttnn.float32, layout=ttnn.TILE_LAYOUT, device=device)

print(ttnn.to_torch(ttnn.expm1_bw(g, t)[0]))
# tensor([[6.7379e-03, 4.5419e-05, 0., 0., 0., 0., 0.]])
print(torch.exp(x.double()))
# tensor([[6.7379e-03, 4.5400e-05, 4.1399e-08, 2.0612e-09, 9.3576e-14, 8.7565e-27, 1.2161e-37]])
```

### Observed vs expected

float32, gradient `1`, golden `exp(x)` in float64. ULP is the integer distance between float32 bit patterns:

| input | got | want | ULP |
|---|---|---|---|
| -5 | 0.006737947 | 0.006737947 | 1 |
| -10 | 4.541874e-05 | 4.539993e-05 | 5,170 |
| -17 | **0** | 4.139938e-08 | 858,902,297 |
| -20 | **0** | 2.061154e-09 | 822,977,587 |
| -30 | **0** | 9.357623e-14 | 701,675,270 |
| -60 | **0** | 8.756511e-27 | 338,522,313 |
| -85 | **0** | 1.216099e-37 | 36,013,792 |

bfloat16, gradient `1`, ULP between bfloat16 bit patterns:

| input | got | want | ULP |
|---|---|---|---|
| -3 | 0.05078125 | 0.04980469 | 4 |
| -4 | 0.01953125 | 0.01831055 | 10 |
| -6.25 | **0** | 0.001930237 | 15,101 |
| -6.28125 | **0** | 0.001869202 | 15,093 |
| -10 | **0** | 4.529953e-05 | 14,398 |
| -40 | **0** | 4.255494e-18 | 8,861 |
| -85 | **0** | 1.219575e-37 | 550 |

The relative error over the zeroed band is 100%. The rows above the threshold show the same cancellation in a milder form: at `x = -10` in float32 the gradient is already 5,170 ULP off.

### Affected input range

All 65,536 bfloat16 bit patterns as the input, gradient `1`, counting inputs whose exact gradient is a normal float but whose result is exactly zero (subnormal inputs and gradients outside the float32 normal range excluded):

| dtype | zeros where the golden is normal | affected range | max ULP | mean ULP |
|---|---|---|---|---|
| bfloat16 | 487 | `[-87, -6.25]` | 15,101 | 105.29 |
| float32 | 295 | `[-87, -17]` | 858,902,297 | 3,266,830 |

### Which architectures

All of them. `expm1_bw` is a host-side composite in `ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp`, so Wormhole and Blackhole build the same op graph and produce the same values.

### How it was verified

ttsim Blackhole v1.10.6 (`TT_METAL_SIMULATOR=libttsim_bh.so`, slow dispatch). Exhaustive sweep of all 65,536 bfloat16 bit patterns as the input with an all-ones gradient, in both bfloat16 and float32, against `exp(x)` evaluated in float64. Subnormal inputs (flushed by the hardware) and gradients outside the float32 normal range are excluded.

`test_backward_expm1.py::test_bw_expm1` does not catch it: it draws inputs from `[-10, 10]` and asserts PCC only, which the large positive values dominate.

expm1_bw before/after

### PR DESCRIPTION
**Pull request by badnikhil**
### PR Category

Bug fix

### Summary

Closes #55723

`expm1_bw` built the derivative as `expm1(x) + 1`. That equals `exp(x)` in exact arithmetic only: `expm1(x)` rounds to exactly `-1` once `exp(x)` falls below half an ULP of `1`, so the `+ 1` cancels and the gradient becomes exactly `0` — for every input below `-6.25` in bfloat16 and `-17` in float32, where the true gradient is still a normal float down to `1.2e-37`.

The derivative of `expm1` is `exp`, so this calls `exp` directly. One op fewer, and no cancellation at all.

All 65,536 bfloat16 bit patterns as the input with an all-ones gradient, golden `exp(x)` in float64, on ttsim Blackhole:

| dtype | zeros where the golden is normal | max ULP | mean ULP |
|---|---|---|---|
| bfloat16 | 487 → **0** | 15,101 → **1** | 105.2871 → **0.0233** |
| float32 | 295 → **0** | 858,902,297 → **1** | 3,266,830.02 → **0.0032** |

Spot values, float32:

| input | before | after | exact |
|---|---|---|---|
| -5 | 0.006737947 | 0.006737947 | 0.006737947 |
| -10 | 4.541874e-05 | **4.539993e-05** | 4.539993e-05 |
| -17 | 0 | **4.139937e-08** | 4.139938e-08 |
| -30 | 0 | **9.357623e-14** | 9.357623e-14 |
| -85 | 0 | **1.216099e-37** | 1.216099e-37 |

expm1_bw before/after

### Notes for reviewers

- Host composite, so both Wormhole and Blackhole are covered by the one change; there is no arch-specific kernel to keep in sync.
- The op graph goes from `expm1 -> add -> multiply` to `exp -> multiply`, so this is one dispatch cheaper as well as exact. `ttnn::exp(input, false, ...)` selects the accurate exp, matching what `ttnn.exp` does by default.
- Positive inputs are unaffected in kind but do get more accurate: `expm1(x) + 1` costs a rounding that `exp(x)` does not, which is where the mean ULP improvement above `x = 0` comes from.
- Edge cases: `exp(x)` overflows to `+inf` above `x = 88.7` exactly as `expm1(x) + 1` did, `exp(-inf)` is `0`, and NaN behaviour is whatever `ttnn.exp` does — none of that changes.
- `test_bw_expm1_all_bitpatterns` sweeps every bfloat16 bit pattern in both bfloat16 and float32, asserts no normal gradient comes back as zero, and holds 2 ULP. It fails both parametrisations on `main` ("487 inputs returned a zero gradient, first at x=-6.25" / "295 ... first at x=-17.0"). The reference is evaluated in float64 rather than through the registered golden, so the assertion is not limited by the reference's own float32 rounding.
- The pre-existing `test_bw_expm1` PCC cases still pass; they did not catch this because they draw from `[-10, 10]` and PCC there is dominated by the large positive values.
- Verified on ttsim Blackhole v1.10.6 (slow dispatch); no silicon.

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp [hunk 1]
```diff
@@ -1331,13 +1331,16 @@ std::vector<Tensor> exp2_bw(
     return grad_tensor;
 }
 
-// bw(expm1) = grad * expm1(input) + 1
+// bw(expm1) = grad * exp(input)
+// expm1(x) + 1 is exp(x) in exact arithmetic only: expm1(x) rounds to exactly -1 once
```
Comment (mateusznowakTT) [inline ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp]: Please remove this incode comment describing the old behavior. PR description is sufficient for this.

File: tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_expm1.py [hunk 2]
```diff
@@ -27,3 +28,50 @@ def test_bw_expm1(input_shapes, device):
...
+    golden = golden64.to(dtype)
+
+    tiny = torch.finfo(torch.float32).tiny
+    finite_in = torch.isfinite(x) & ((x == 0) | (x.abs() >= tiny))
+    normal_out = (golden64.abs() >= tiny) & (golden64.abs() <= torch.finfo(torch.float32).max)
+    checked = finite_in & normal_out
+
+    lost = checked & (result == 0)
+    assert lost.sum() == 0, f"{int(lost.sum())} inputs returned a zero gradient, first at x={float(x[lost][0])}"
+    assert_with_ulp(golden[checked], result[checked], ulp_threshold=2)
```
Comment (mateusznowakTT) [inline tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_expm1.py]: assign_with_ulp is being changed with https://github.com/tenstorrent/tt-metal/pull/55661/ - please align after it will be merged.

File: tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_expm1.py [hunk 3]
```diff
@@ -27,3 +28,50 @@ def test_bw_expm1(input_shapes, device):
 
     comp_pass = compare_pcc(tt_output_tensor_on_device, golden_tensor)
     assert comp_pass
+
+
+@pytest.mark.parametrize("dtype", [torch.bfloat16, torch.float32])
+def test_bw_expm1_all_bitpatterns(device, dtype):
+    """Exhaustive: every bfloat16 bit pattern as the input, with an all-ones gradient.
+
+    d/dx expm1(x) is exp(x), which is a normal float down to x = -87.34, so no finite input
```
Comment (mateusznowakTT) [inline tests/ttnn/nightly/unit_tests/operations/eltwise/backward/test_backward_expm1.py:37]: Old behavior description is unnecessary here.

**Review by mateusznowakTT — CHANGES_REQUESTED** [review]

**Comment by badnikhil** [comment]
Both old-behaviour comments are gone, and assert_with_ulp now uses the keyword form from #55661. 85b805f1

**Review by mateusznowakTT — APPROVED** [review]
